pilot/pkg/bootstrap/server_test.go

```diff
@@ -627,6 +630,57 @@ func TestInitOIDC(t *testing.T) {
 	}
 }
 
+func TestWatchDNSCertForK8sCA(t *testing.T) {
+	tests := []struct {
+		name        string
+		certToWatch []byte
+		certRotated bool
+	}{
+		{
+			name:        "expired cert rotation",
+			certToWatch: testcerts.ExpiredServerCert,
+			certRotated: true,
+		},
+		{
+			name:        "valid cert no rotation",
+			certToWatch: testcerts.ServerCert,
+			certRotated: false,
+		},
+	}
+
+	csr := &cert.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "abc.xyz",
+		},
+		Status: cert.CertificateSigningRequestStatus{
+			Certificate: testcerts.ServerCert,
+		},
+	}
+	s := &Server{
+		server:                  server.New(),
+		istiodCertBundleWatcher: keycertbundle.NewWatcher(),
+		kubeClient:              kube.NewFakeClient(csr),
+		dnsNames:                []string{"abc.xyz"},
+	}
+
+	for _, tt := range tests {
+		stop := make(chan struct{})
+		t.Run(tt.name, func(t *testing.T) {
+			s.istiodCertBundleWatcher.SetAndNotify(testcerts.ServerKey, tt.certToWatch, testcerts.CACert)
+
+			go s.watchDNSCertForK8sCA(2*time.Second, stop, "", "test-signer", true, time.Duration(0))
+			sleep.Until(stop, 10*time.Second)
```

Comment:
sleep() is not appropriate in tests, see https://github.com/istio/istio/wiki/test%20flakes

Reply:
IMO: one way is to keep periodically polling the newer cert to compare with the older cert to see if there is a diff and report an error if they are the same after timeout

Reply:
Now we are not having any specific timeInterval for sleep. We are using already expired cert so watchDNSCertForK8sCA should not sleep and execute cert rotation with fake client. I am adding sleep of 500 ms in test before sending stop signal to watchDNSCertForK8sCA to let watchDNSCertForK8sCA complete once.

Reply:
I have added sleep of 1 sec in test to wait for watchDNSCertForK8sCA to complete before we send stop signal, I agree it might be flaky sometimes.
@howardjohn @shankgan  We can do few things :-
1. Remove 'valid cert no rotation' test scenario from unit test and add retry untill cert changes with 10sec timeout. It will work for 'expired cert rotation' scenario as cert will be changed.
2. Take out whatever is inside for loop in watchDNSCertForK8sCA and create a new function for that. Now test this new function for unit testing for both test scenarios as this new function will run only once if called directly. Combine this with 1st point to test watchDNSCertForK8sCA for expired cert rotation using retryUntill.
3. Let it be, observe if test is flaky and fix if test is flaky in future.

Reply:
"expired cert rotation" appears to be the flaky test with a 1 sec timeout. IMO, we could add two separate timeouts - 10 sec for expired cert rotation and 1 sec for  'valid cert no rotation'. In former case we can use retryUntil to keep polling until cert has rotated. In later case, it makes sense to sleep for 1 sec and just check that cert has not rotated.

Reply:
@howardjohn @shankgan 
Added retryUntil for 10sec timeout with adding sleep for 1sec in starting of validation of cert changed.

1. For Expired cert case

- It will retry in every 1 second for total 10 seconds. In most cases cert should be changed in 1 sec so test will not take much time. 10 second timeout minimise test flakiness.

2. For valid cert no rotation case

- After 1st run (in one second) the flow will come out of RetryUntill block as cert is same, so this should complete in one second.